browser/ui/views/text_recognition_dialog_view.cc

```diff
@@ -0,0 +1,183 @@
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/views/text_recognition_dialog_view.h"
+
+#include <memory>
+
+#include "base/bind.h"
+#include "base/strings/string_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/task/task_traits.h"
+#include "base/task/thread_pool.h"
+#include "brave/components/l10n/common/localization_util.h"
+#include "brave/components/text_recognition/browser/text_recognition.h"
+#include "brave/grit/brave_generated_resources.h"
+#include "components/constrained_window/constrained_window_views.h"
+#include "content/public/browser/web_contents_user_data.h"
+#include "ui/base/clipboard/scoped_clipboard_writer.h"
+#include "ui/base/metadata/metadata_impl_macros.h"
+#include "ui/gfx/geometry/insets.h"
+#include "ui/views/controls/label.h"
+#include "ui/views/controls/scroll_view.h"
+#include "ui/views/layout/flex_layout.h"
+#include "ui/views/widget/widget.h"
+#include "ui/views/widget/widget_observer.h"
+#include "ui/views/window/dialog_client_view.h"
+
+namespace {
+
+// Tracks whether text recognition dialog is active or not for WebContents.
+class TextRecognitionDialogTracker
+    : public content::WebContentsUserData<TextRecognitionDialogTracker>,
+      public views::WidgetObserver {
+ public:
+  TextRecognitionDialogTracker(const TextRecognitionDialogTracker&) = delete;
+  TextRecognitionDialogTracker& operator=(const TextRecognitionDialogTracker&) =
+      delete;
+  ~TextRecognitionDialogTracker() override = default;
+
+  void SetActiveDialog(views::Widget* widget) {
+    DCHECK(!active_dialog_);
+    active_dialog_ = widget;
+    active_dialog_->AddObserver(this);
+  }
+
+  views::Widget* active_dialog() { return active_dialog_; }
+
+ private:
+  friend class content::WebContentsUserData<TextRecognitionDialogTracker>;
+  explicit TextRecognitionDialogTracker(content::WebContents* web_contents)
+      : content::WebContentsUserData<TextRecognitionDialogTracker>(
+            *web_contents) {}
+
+  // views::WidgetObserver overrides
+  void OnWidgetClosing(views::Widget* widget) override {
+    DCHECK_EQ(active_dialog_, widget);
+    DCHECK(active_dialog_->HasObserver(this));
+    active_dialog_->RemoveObserver(this);
+    active_dialog_ = nullptr;
+  }
+
+  raw_ptr<views::Widget> active_dialog_ = nullptr;
+
+  WEB_CONTENTS_USER_DATA_KEY_DECL();
+};
+
+WEB_CONTENTS_USER_DATA_KEY_IMPL(TextRecognitionDialogTracker);
+
+}  // namespace
+
+namespace brave {
+
+void ShowTextRecognitionDialog(content::WebContents* web_contents,
+                               const SkBitmap& image) {
+  // Use existing dialog instead of creating multiple dialog.
+  // Dialog will have lastly recognizied text from image from same tab.
+  TextRecognitionDialogTracker::CreateForWebContents(web_contents);
+  auto* dialog_tracker =
+      TextRecognitionDialogTracker::FromWebContents(web_contents);
+
+  if (auto* active_dialog = dialog_tracker->active_dialog()) {
```

Comment:
will this work when you close the dialog and request the recognition again on the same page?

Reply:
Yes, `dialog_tracker->active_dialog()` will give `nullptr` when dialog is closed.

Reply:
oh, I missed `OnWidgetClosing` handler. that's great then!

Reply:
I'd like to suggest OnWidgetClosing be replaced with OnWidgetDestroying or Destroyed. The API isn't always called based on how it's getting started to close, so it's considered as undesirable, I think.

https://source.chromium.org/chromium/chromium/src/+/main:ui/views/widget/widget_observer.h;l=25;drc=0245f5dd99fb2af653ac5a231395d70054020c39;bpv=0;bpt=1

Reply:
@sangwoo108 Updated. Thanks! 👍🏼